docs/Implementing-easyblocks.rst

```diff
@@ -0,0 +1,486 @@
+* consists only of lower-case alphanumeric characters (``[a-z0-9]``) and underscores (``_``)
+
+  * dashes (``-``) are replaced by underscores (``_``)
+  * any other non-alphanumeric characters (incl. spaces) are simply dropped
+
+Examples include:
+
+* ``gcc.py`` (for *GCC*)
+* ``netcdf_fortran.py`` (for *netCDF-Fortran*)
+* ``gamess_us.py`` (for *GAMESS (US)*)
+
+The ``get_module_path`` function provided in ``easybuild.framework.easyconfig.easyconfig`` returns the (full)
+module location for a particular software name or easyblock class name; for example:
+
+.. code::
+
+  >>> from easybuild.framework.easyconfig.easyconfig import get_module_path
+  >>> get_module_path('netCDF-Fortran')
+  'easybuild.easyblocks.netcdf_fortran'
+  >>> get_module_path('EB_netCDF_minus_Fortran')
+  'easybuild.easyblocks.netcdf_fortran'
+
+The location of the Python module is determined by whether the easyblock is generic or software-specific.
+Generic easyblocks are located in the ``easybuid.easyblocks.generic`` namespace, while software-specific easyblocks
+live in the ``easybuild.easyblocks`` namespace directly. To keep things organised, the actual Python module file
+for software-specific easyblocks are kept in 'letter' subdirectories,
+rather than in one large '``easyblocks``' directory
+(see https://github.com/hpcugent/easybuild-easyblocks/blob/master/easybuild/easyblocks/).
+
+Note that you shouldn't concern yourself too much with getting the location of an easyblock right, as long as you
+use ``--include-easyblocks`` to make EasyBuild use additional or customised easyblocks;
+see :ref:`include_easyblocks` for more information.
+
+
+.. _implementing_easyblocks_structure:
+
+Structure of an easyblock
+-------------------------
+
+The example below shows the overal structure of an easyblock:
+
+.. code:: python
+
+    from easybuild.framework.easyblock import EasyBlock
+    from easybuild.tools.filetools import run_cmd
+
+    class EB_Example(EasyBlock):
+        """Custom easyblock for Example"""
+
+        def configure_step(self):
+            """Custom implementation of configure step for Example"""
+
+            # run configure.sh to configure the build
+            run_cmd("./configure.sh PREFIX=%s" % self.installdir)
+
+
+Each easyblock includes an implementation of a ``class`` that (directly or indirectly) derives from the abstract
+``EasyBlock`` class.
+
+Typically some useful functions provided by the EasyBuild framework are imported at the top of the Python module.
+
+In the class definition, one or more '``*_step``' methods are redefined, to implement the corresponding step
+in the build and installation procedure.
+
+Each easyblock *must* implement the ``configure``, ``build`` and ``install`` steps, since these are not implemented
+in the abstract ``EasyBlock`` class. This could be done explicitely by redefining the corresponding ``*_step`` methods,
+or implicitely by :ref:`implementing_easyblocks_deriving`.
```

Comment:
It reads weird to have a capital letter in the middle of the sentence just to take advantage of the link. Given that it links to the next line...... maybe it makes sense to remove it?